File: src/bombe/indexer/semantic.py
"""Optional semantic hint integrations for call resolution."""

from __future__ import annotations

import json
import os
import shutil
from pathlib import Path
from typing import Any


def backend_statuses() -> list[dict[str, Any]]:
    backends = [
        ("pyright", "pyright"),
        ("typescript_language_server", "typescript-language-server"),
        ("gopls", "gopls"),
        ("jdtls", "jdtls"),
    ]
    payload: list[dict[str, Any]] = []
    for name, executable in backends:
        location = shutil.which(executable)
        payload.append(
            {
                "backend": name,
                "available": location is not None,
                "executable": location,
            }
        )
    return payload


def _normalize_relative_path(relative_path: str) -> str:
    return relative_path.strip().lstrip("/").replace("\\", "/")


def _sidecar_path(repo_root: Path, relative_path: str) -> Path:
    normalized = _normalize_relative_path(relative_path)
    return repo_root / ".bombe" / "semantic" / f"{normalized}.hints.json"


def _load_json(path: Path) -> dict[str, Any] | None:
    if not path.exists():
        return None
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return None
    if not isinstance(payload, dict):
        return None
    return payload


def _merge_hint_entry(
    target: dict[tuple[int, str], set[str]],
    receiver: str,
    owner_type: str,
    start_line: int,
    end_line: int,
) -> None:
    normalized_receiver = receiver.strip()
    normalized_owner = owner_type.strip()
    if not normalized_receiver or not normalized_owner:
        return
    start = max(1, start_line)
    end = max(start, end_line)
    for line in range(start, min(end, start + 512) + 1):
        target.setdefault((line, normalized_receiver), set()).add(normalized_owner)


def _parse_hint_payload(payload: dict[str, Any]) -> dict[tuple[int, str], set[str]]:
    hints: dict[tuple[int, str], set[str]] = {}
    entries = payload.get("receiver_hints", [])
    if not isinstance(entries, list):
        return hints
    for item in entries:
        if not isinstance(item, dict):
            continue
        receiver = str(item.get("receiver", "")).strip()
        owner_type = str(item.get("owner_type", "")).strip()
        if not receiver or not owner_type:
            continue
        line = item.get("line")
        line_start = item.get("line_start", line)
        line_end = item.get("line_end", line)
        try:
            start = int(line_start if line_start is not None else 1)
            end = int(line_end if line_end is not None else start)
        except (TypeError, ValueError):
            continue
        _merge_hint_entry(hints, receiver, owner_type, start, end)
    return hints


def _merge_hint_maps(
    target: dict[tuple[int, str], set[str]],
    source: dict[tuple[int, str], set[str]],
) -> dict[tuple[int, str], set[str]]:
    for key, values in source.items():
        target.setdefault(key, set()).update(values)
    return target


def load_receiver_type_hints(
    repo_root: Path,
    relative_path: str,
) -> dict[tuple[int, str], set[str]]:
    hints: dict[tuple[int, str], set[str]] = {}
    normalized_relative_path = _normalize_relative_path(relative_path)

    sidecar = _sidecar_path(repo_root, normalized_relative_path)
    sidecar_payload = _load_json(sidecar)
    if sidecar_payload is not None:
        _merge_hint_maps(hints, _parse_hint_payload(sidecar_payload))

    global_hints_file = os.getenv("BOMBE_SEMANTIC_HINTS_FILE", "").strip()
    if global_hints_file:
        payload = _load_json(Path(global_hints_file).expanduser().resolve())
        if payload is not None:
            files = payload.get("files", {})
            if isinstance(files, dict):
                lookup_candidates = {
                    normalized_relative_path,
                    relative_path,
                    relative_path.replace("\\", "/").lstrip("/"),
                }
                for candidate in lookup_candidates:
                    file_payload = files.get(candidate)
                    if isinstance(file_payload, dict):
                        _merge_hint_maps(hints, _parse_hint_payload(file_payload))
    return hints

File: crates/bombe-core/src/indexer/semantic.rs
//! Optional semantic hint integrations for call resolution.

use std::collections::{HashMap, HashSet};
use std::path::Path;

/// Load receiver type hints from sidecar JSON files and environment.
pub fn load_receiver_type_hints(
    repo_root: &Path,
    relative_path: &str,
) -> HashMap<(i64, String), HashSet<String>> {
    let mut hints: HashMap<(i64, String), HashSet<String>> = HashMap::new();
    let normalized = normalize_relative_path(relative_path);

    // Load sidecar hints
    let sidecar = repo_root
        .join(".bombe")
        .join("semantic")
        .join(format!("{normalized}.hints.json"));
    if let Some(payload) = load_json(&sidecar) {
        merge_hint_maps(&mut hints, &parse_hint_payload(&payload));
    }

    // Load global hints file
    if let Ok(global_path) = std::env::var("BOMBE_SEMANTIC_HINTS_FILE") {
        let global_path = global_path.trim().to_string();
        if !global_path.is_empty() {
            let expanded = Path::new(&global_path);
            if let Some(payload) = load_json(expanded) {
                if let Some(files) = payload.get("files").and_then(|v| v.as_object()) {
                    let candidates = [
                        normalized.clone(),
                        relative_path.to_string(),
                        relative_path
                            .replace('\\', "/")
                            .trim_start_matches('/')
                            .to_string(),
                    ];
                    for candidate in &candidates {
                        if let Some(file_payload) = files.get(candidate) {
                            if let Some(obj) = file_payload.as_object() {
                                let val = serde_json::Value::Object(obj.clone());
                                merge_hint_maps(&mut hints, &parse_hint_payload(&val));
                            }
                        }
                    }
                }
            }
        }
    }

    hints
}

fn normalize_relative_path(path: &str) -> String {
    path.trim().trim_start_matches('/').replace('\\', "/")
}

fn load_json(path: &Path) -> Option<serde_json::Value> {
    if !path.exists() {
        return None;
    }
    let content = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&content).ok()
}

fn parse_hint_payload(payload: &serde_json::Value) -> HashMap<(i64, String), HashSet<String>> {
    let mut hints: HashMap<(i64, String), HashSet<String>> = HashMap::new();
    let entries = match payload.get("receiver_hints").and_then(|v| v.as_array()) {
        Some(arr) => arr,
        None => return hints,
    };
    for item in entries {
        let obj = match item.as_object() {
            Some(o) => o,
            None => continue,
        };
        let receiver = obj
            .get("receiver")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        let owner_type = obj
            .get("owner_type")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if receiver.is_empty() || owner_type.is_empty() {
            continue;
        }
        let line = obj.get("line").and_then(|v| v.as_i64());
        let line_start = obj
            .get("line_start")
            .and_then(|v| v.as_i64())
            .or(line)
            .unwrap_or(1);
        let line_end = obj
            .get("line_end")
            .and_then(|v| v.as_i64())
            .unwrap_or(line_start);
        let start = line_start.max(1);
        let end = line_end.max(start);
        for line_num in start..=(end.min(start + 512)) {
            hints
                .entry((line_num, receiver.clone()))
                .or_default()
                .insert(owner_type.clone());
        }
    }
    hints
}

fn merge_hint_maps(
    target: &mut HashMap<(i64, String), HashSet<String>>,
    source: &HashMap<(i64, String), HashSet<String>>,
) {
    for (key, values) in source {
        target
            .entry(key.clone())
            .or_default()
            .extend(values.iter().cloned());
    }
}
